runtime/near-vm-logic/src/logic.rs

```diff
@@ -1204,6 +1206,94 @@ impl<'a> VMLogic<'a> {
         self.gas_counter.pay_wasm_gas(opcodes)
     }
 
+    /// Verify aggregate BLS12-381 signature for given message and
+    /// public keys list (or one aggregate public key)
+    ///
+    /// The length of each public key(include aggregate one) should be 48 bytes and
+    /// the length of the signature is 96 bytes.
+    ///
+    /// In case of providing the list of public keys, all public keys should be written
+    /// in one slice one by one without any separation. `pubkeys_ptr` - the pointer
+    /// to the beginning of this slice and `pubkeys_len` the len of this
+    /// slice in bytes (=48 * PUBLIC_KEYS_CNT).
+    ///
+    /// Returns 1 if the signature verification pass, and 0 if verification fail.
+    ///
+    /// # Errors
+    ///
+    /// If the input data is incorrect returns [HostError::Bls1238VerifyError]
+    ///
+    /// # Cost
+    ///
+    /// `input_cost(aggregate_signature_len) + input_cost(msg_len) + input_cost(pubkey_len) +
+    ///  bls12381_verify_base + bls12381_verify_byte * msg_len + bls12381_verify_elements * pubkeys_cnt`
+    ///
+    /// Where `pubkeys_cnt` -- the number of public keys (`pubkeys_len/48`).
+    ///
+    /// The signature_len is a const = 96 bytes. And for signature verification, first,
+    /// the public keys are aggregated into one public key and after that the signature is verified,
+    /// so public keys number and message len influence on cost independently.
+    pub fn bls12_381_aggregate_verify(
+        &mut self,
+        aggregate_signature_ptr: u64,
+        aggregate_signature_len: u64,
+        msg_ptr: u64,
+        msg_len: u64,
+        pubkeys_ptr: u64,
+        pubkeys_len: u64,
+    ) -> Result<u64> {
+        self.gas_counter.pay_base(bls12381_verify_base)?;
+        const PUBKEY_LEN: u64 = 48;
+
+        let aggregate_signature =
+            self.get_vec_from_memory_or_register(aggregate_signature_ptr, aggregate_signature_len)?;
+        let message = self.get_vec_from_memory_or_register(msg_ptr, msg_len)?;
+        let pubkeys_raw = self.get_vec_from_memory_or_register(pubkeys_ptr, pubkeys_len)?;
+        let num_pubkeys = (pubkeys_raw.len() as u64) / PUBKEY_LEN;
```

Comment:
is there a need to add a check in case pubkeys raw length isn't a multiple of `PUBKEY_LEN`?

Reply:
fixed: https://github.com/near/nearcore/pull/8184/commits/6355c63184f168c73c1ca4edca02fc732b8abf0c